pkg/webhook/bootstrap.go

```diff
@@ -0,0 +1,309 @@
+}
+
+func (s *Server) clientConfig() (*webhookClientConfig, error) {
+	if s.Host != nil && s.Service != nil {
+		return nil, errors.New("URL and Service can't be set at the same time")
+	}
+	cc := &webhookClientConfig{
+		WebhookClientConfig: &admissionregistration.WebhookClientConfig{
+			CABundle: []byte{},
+		},
+	}
+	if s.Host != nil {
+		u := url.URL{
+			Scheme: "https",
+			Host:   *s.Host,
+		}
+		urlString := u.String()
+		cc.URL = &urlString
+	}
+	if s.Service != nil {
+		cc.Service = &admissionregistration.ServiceReference{
+			Name:      s.Service.Name,
+			Namespace: s.Service.Namespace,
+			// Path will be set later
+		}
+	}
+	return cc, nil
+}
+
+func (w *webhookClientConfig) withPath(path string) error {
+	if w.URL != nil {
+		u, err := url.Parse(*w.URL)
+		if err != nil {
+			return err
+		}
+		u.Path = path
+		urlString := u.String()
+		w.URL = &urlString
+	}
+	if w.Service != nil {
+		w.Service.Path = &path
+	}
+	return nil
+}
+
+// whConfigs creates a mutatingWebhookConfiguration and(or) a validatingWebhookConfiguration based on registry.
+// For the same type of webhook configuration, it generates a webhook entry per endpoint.
+func (s *Server) whConfigs() ([]runtime.Object, error) {
+	mutatingWebhooks, validatingWebhooks := []v1beta1.Webhook{}, []v1beta1.Webhook{}
+	var mutatingWebhookConfig *admissionregistration.MutatingWebhookConfiguration
+	var validatingWebhookConfig *admissionregistration.ValidatingWebhookConfiguration
+	for path, webhook := range s.registry {
+		var admissionWebhook *admission.Webhook
+		if webhook.GetType() != types.WebhookTypeMutating && webhook.GetType() != types.WebhookTypeValidating {
+			continue
+		}
+
+		admissionWebhook = webhook.(*admission.Webhook)
+		wh := admissionregistration.Webhook{
+			Name:              admissionWebhook.GetName(),
+			Rules:             admissionWebhook.Rules,
+			FailurePolicy:     admissionWebhook.FailurePolicy,
+			NamespaceSelector: admissionWebhook.NamespaceSelector,
+			ClientConfig: admissionregistration.WebhookClientConfig{
+				// The reason why we assign an empty byte array to CABundle is that
+				// CABundle field will be updated by the Provisioner.
+				CABundle: []byte{},
+			},
+		}
+		cc, err := s.clientConfig()
+		if err != nil {
+			return nil, err
+		}
+		err = cc.withPath(path)
+		if err != nil {
+			return nil, err
+		}
+
+		switch webhook.GetType() {
+		case types.WebhookTypeMutating:
+			mutatingWebhooks = append(mutatingWebhooks, wh)
+		case types.WebhookTypeValidating:
+			validatingWebhooks = append(validatingWebhooks, wh)
+		}
+	}
+
+	objects := []runtime.Object{}
+	if len(mutatingWebhooks) > 0 {
+		mutatingWebhookConfig = &admissionregistration.MutatingWebhookConfiguration{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "admissionregistration.k8s.io/v1beta1",
+				Kind:       "MutatingWebhookConfiguration",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        "controller-runtime-webhooks",
```

Comment:
If two controller-runtime managers are running, will they fight over these objects? Might be a good idea to make these names configurable.